autogluon/utils/tabular/features/generators/pipeline.py

```diff
@@ -0,0 +1,113 @@
+import copy
+import logging
+
+import psutil
+from pandas import DataFrame
+
+from .bulk import BulkFeatureGenerator
+from .dummy import DummyFeatureGenerator
+from .drop_unique import DropUniqueFeatureGenerator
+from .fillna import FillNaFeatureGenerator
+from ..feature_metadata import FeatureMetadata
+from ..types import get_type_map_real
+from ...data.utils import get_approximate_df_mem_usage
+
+logger = logging.getLogger(__name__)
+
+
+# TODO: Documentation
+class PipelineFeatureGenerator(BulkFeatureGenerator):
+    """
+    PipelineFeatureGenerator is an implementation of BulkFeatureGenerator with various smart defaults and edge case handling functionality to enable robust data handling.
+    It is recommended that users base any custom feature generators meant for end-to-end data transformation from PipelineFeatureGenerator.
+        Reference AutoMLPipelineFeatureGenerator for an example of extending PipelineFeatureGenerator.
+    It is not recommended that PipelineFeatureGenerator be used as a generator within any other generator's pre or post generators.
+    """
+    def __init__(self, pre_generators=None, post_generators=None, pre_drop_useless=True, pre_enforce_types=True, reset_index=True, verbosity=3, **kwargs):
+        if pre_generators is None:
+            pre_generators = [FillNaFeatureGenerator(inplace=True)]
+        if post_generators is None:
+            post_generators = [DropUniqueFeatureGenerator()]
+
+        super().__init__(pre_generators=pre_generators, post_generators=post_generators, pre_drop_useless=pre_drop_useless, pre_enforce_types=pre_enforce_types, reset_index=reset_index, verbosity=verbosity, **kwargs)
+
+        self._feature_metadata_in_real: FeatureMetadata = None  # FeatureMetadata object based on the original input features real dtypes (will contain dtypes such as 'int16' and 'float32' instead of 'int' and 'float').
+
+        self._is_dummy = False  # If True, returns a single dummy feature as output. Occurs if fit with no useful features.
+
+        self.pre_memory_usage = None
+        self.pre_memory_usage_per_row = None
+        self.post_memory_usage = None
+        self.post_memory_usage_per_row = None
+
+    def fit_transform(self, X: DataFrame, y=None, feature_metadata_in: FeatureMetadata = None, **kwargs) -> DataFrame:
+        self._ensure_no_duplicate_column_names(X=X)  # TODO: Remove this, move pre_memory_usage and post_memory_usage into super().
+        self._compute_pre_memory_usage(X)
+        X_out = super().fit_transform(X=X, y=y, feature_metadata_in=feature_metadata_in, **kwargs)
+        self._compute_post_memory_usage(X_out)
```

Comment:
Since this is supposed to handle edge cases, should there be some final check of `X_out` + warning if user's custom `_transform` creates features that should've been removed?  Eg. duplicate or constant features.  I imagine this could easily happen with poorly created custom generators (or in future HPO applied to generator-configurations)

Reply:
Added TODO